File: fsharp/protein-translation/ProteinTranslation.fs
module ProteinTranslation

    type private RNASequence =
        | Methionine
        | Phenylalanine
        | Leucine
        | Serine
        | Tyrosine
        | Cysteine
        | Tryptophan
        | STOP

    let private RNATranslation (codon:char array) =
        match codon with
        | [|'A'; 'U'; 'G'|]
            -> Methionine
        | [|'U'; 'U'; 'U'|]
        | [|'U'; 'U'; 'C'|]
            -> Phenylalanine
        | [|'U'; 'U'; 'A'|]
        | [|'U'; 'U'; 'G'|]
            -> Leucine
        | [|'U'; 'C'; 'U'|]
        | [|'U'; 'C'; 'C'|]
        | [|'U'; 'C'; 'A'|]
        | [|'U'; 'C'; 'G'|]
            -> Serine
        | [|'U'; 'A'; 'U'|]
        | [|'U'; 'A'; 'C'|]
            -> Tyrosine
        | [|'U'; 'G'; 'U'|]
        | [|'U'; 'G'; 'C'|]
            -> Cysteine
        | [|'U'; 'G'; 'G'|]
            -> Tryptophan
        | [|'U'; 'A'; 'A'|]
        | [|'U'; 'A'; 'G'|]
        | [|'U'; 'G'; 'A'|]
            -> STOP
        | _ -> failwith "Not a valid codon."

    let private toString (rnaSequence:RNASequence):string =
        match rnaSequence with
            | Methionine -> "Methionine"
            | Phenylalanine -> "Phenylalanine"
            | Leucine -> "Leucine"
            | Serine -> "Serine"
            | Tyrosine -> "Tyrosine"
            | Cysteine -> "Cysteine"
            | Tryptophan -> "Tryptophan"
            | STOP -> ""

    let proteins (rna:string):string list =
        rna.ToCharArray()
        |> Array.chunkBySize 3
        |> List.ofArray
        |> List.map RNATranslation
        |> List.takeWhile ((<>) STOP)
        |> List.map toString

File: solutions/scala/protein-translation/1/src/main/scala/ProteinTranslation.scala
enum RNASequence
{
    case Methionine
    case Phenylalanine
    case Leucine
    case Serine
    case Tyrosine
    case Cysteine
    case Tryptophan
    case STOP
}

object ProteinTranslation
{

    private def RNATranslation (codon:Array[Char]):RNASequence =
        codon match
        case Array('A', 'U', 'G')
            => RNASequence.Methionine
        case Array('U', 'U', 'U')
        | Array('U', 'U', 'C')
            => RNASequence.Phenylalanine
        case Array('U', 'U', 'A')
        | Array('U', 'U', 'G')
            => RNASequence.Leucine
        case Array('U', 'C', 'U')
        | Array('U', 'C', 'C')
        | Array('U', 'C', 'A')
        | Array('U', 'C', 'G')
            => RNASequence.Serine
        case Array('U', 'A', 'U')
        | Array('U', 'A', 'C')
            => RNASequence.Tyrosine
        case Array('U', 'G', 'U')
        | Array('U', 'G', 'C')
            => RNASequence.Cysteine
        case Array('U', 'G', 'G')
            => RNASequence.Tryptophan
        case Array('U', 'A', 'A')
        | Array('U', 'A', 'G')
        | Array('U', 'G', 'A')
            => RNASequence.STOP

    private def toString (rnaSequence:RNASequence):String =
        rnaSequence match
        case RNASequence.Methionine => "Methionine"
        case RNASequence.Phenylalanine => "Phenylalanine"
        case RNASequence.Leucine => "Leucine"
        case RNASequence.Serine => "Serine"
        case RNASequence.Tyrosine => "Tyrosine"
        case RNASequence.Cysteine => "Cysteine"
        case RNASequence.Tryptophan => "Tryptophan"
        case RNASequence.STOP => ""

    def proteins (rna:String):Seq[String] = rna.toCharArray().grouped(3).map(RNATranslation).takeWhile((rnaSequence:RNASequence) => rnaSequence != RNASequence.STOP).map(toString).toSeq
}